File: frontend/src/inference/fuzzy.ts
// Motor difuso — membresía trapezoidal/triangular + firing strength mínimo
// by Jose Rodas

function membership(value: number, conjunto: any): number {
  const tipo = conjunto.tipo ?? 'trapezoidal';
  const pts  = (conjunto.puntos ?? []).map(Number);

  if (tipo === 'triangular' && pts.length >= 3) {
    const [a, b, c] = pts;
    if (value <= a || value >= c) return 0;
    return value <= b ? (value - a) / (b - a) : (c - value) / (c - b);
  }

  if (tipo === 'trapezoidal' && pts.length >= 4) {
    const [a, b, c, d] = pts;
    if (value <= a || value >= d) return 0;
    if (value >= b && value <= c) return 1;
    return value < b ? (value - a) / (b - a) : (d - value) / (d - c);
  }

  return 0;
}

function membershipCat(value: string, conjunto: any): number {
  return String(value).toLowerCase() === String(conjunto.valor ?? '').toLowerCase() ? 1 : 0;
}

export function inferirFuzzy(
  hechos: Record<string, any>,
  variables: any[],
  reglas: any[],
): { conclusion: string; certeza: number; firing: number; explicacion: string }[] {
  const varMap = new Map(variables.map((v: any) => [v.nombre, v]));
  const resultados: { conclusion: string; certeza: number; firing: number; explicacion: string }[] = [];

  for (const r of reglas) {
    const grados: number[] = [];

    for (const cond of (r.condiciones ?? [])) {
      const varDef = varMap.get(cond.var);
      if (!varDef) { grados.push(0); continue; }

      const conjunto = (varDef.conjuntos ?? []).find((c: any) => c.nombre === cond.conjunto);
      if (!conjunto) { grados.push(0); continue; }

      const val = hechos[cond.var];
      if (val === undefined || val === '') { grados.push(0); continue; }

      grados.push(
        varDef.tipo === 'numerico'
          ? membership(parseFloat(val), conjunto)
          : membershipCat(String(val), conjunto),
      );
    }

    if (!grados.length) continue;
    const firing  = Math.min(...grados);
    const certeza = Math.round(firing * parseFloat(r.peso ?? 1.0) * 10000) / 10000;
    if (certeza > 0) {
      resultados.push({ conclusion: r.conclusion, certeza, firing: Math.round(firing * 10000) / 10000, explicacion: r.explicacion ?? '' });
    }
  }

  return resultados.sort((a, b) => b.certeza - a.certeza);
}

File: backend/engines/fuzzy.py
import math

def _membership(value: float, conjunto: dict) -> float:
    tipo = conjunto.get("tipo", "trapezoidal")
    pts  = conjunto.get("puntos", [])
    v    = float(value)

    if tipo == "triangular" and len(pts) >= 3:
        a, b, c = float(pts[0]), float(pts[1]), float(pts[2])
        if v <= a or v >= c: return 0.0
        return (v - a) / (b - a) if v <= b else (c - v) / (c - b)

    if tipo == "trapezoidal" and len(pts) >= 4:
        a, b, c, d = float(pts[0]), float(pts[1]), float(pts[2]), float(pts[3])
        if v <= a or v >= d: return 0.0
        if b <= v <= c:      return 1.0
        return (v - a) / (b - a) if v < b else (d - v) / (d - c)

    return 0.0


def _membership_cat(value: str, conjunto: dict) -> float:
    return 1.0 if str(value).lower() == str(conjunto.get("valor", "")).lower() else 0.0


def inferir_difuso(hechos: dict, variables: list[dict], reglas: list[dict]) -> list[dict]:
    var_map = {v["nombre"]: v for v in variables}
    resultados = []

    for r in reglas:
        grados = []
        for cond in r.get("condiciones", []):
            var_def = var_map.get(cond["var"])
            if not var_def:
                grados.append(0.0)
                continue
            conjunto = next((c for c in var_def.get("conjuntos", []) if c["nombre"] == cond["conjunto"]), None)
            if not conjunto:
                grados.append(0.0)
                continue
            val = hechos.get(cond["var"])
            if val is None:
                grados.append(0.0)
                continue
            if var_def["tipo"] == "numerico":
                grados.append(_membership(float(val), conjunto))
            else:
                grados.append(_membership_cat(str(val), conjunto))

        if not grados:
            continue
        firing = min(grados)
        certeza = round(firing * float(r.get("peso", 1.0)), 4)
        if certeza > 0:
            resultados.append({
                "conclusion":  r["conclusion"],
                "certeza":     certeza,
                "firing":      round(firing, 4),
                "explicacion": r.get("explicacion", ""),
            })

    return sorted(resultados, key=lambda x: -x["certeza"])
